### PR TITLE
xds: refactor resource subscription implementation in XdsClient

### PR DESCRIPTION
**Pull request by voidzcy**
This would be a great maintainability improvement for XdsClientImpl.

Should have no behavior change.

Currently for each type of resources, there are multiple pieces: watchers, cached entries, initial fetch timer, etc. Every time newly subscribing to a new resource or receiving a response, these pieces of state need to be updated individually. This is complicated to maintain and there is a lot of duplicated code. This PR groups status of a subscribed resource into `ResourceSubscriber`. You can think of it for tracking the state of a single resource.

Every time newly subscribing to some resource, a corresponding `ResourceSubscriber` is created. Note it does not control the resource discovery RPCs. It is still the XdsClient that sends RPCs for with all subscribed resource names for each type. A `ResourceSubscriber` can have the following states:

- When the initial resource fetch timer (`respTimer`) is pending, the resource is under discovery, the resource data is unknown. Even if the XdsClient receives a response not containing the corresponding resource, it does not mean the resource is absent. We still need to wait until a response containing the resource data coming or the timer being fired. The timer is scheduled when the `ResourceSubscriber` is created. So the XdsClient should always create the corresponding `ResourceSubscriber` when it starts to subscribe a new resource.

- If the resource fetch timer is not pending, we must know the existence of the resource data. If `data` field is set, it is the most recently received resource data (aka, cached entry). Otherwise, `absent` field is set to `true`, indicating the resource does not exist.

- The exceptional case is when the ADS stream is closed and in the retry backoff period. During that period, `respTimer` is cancelled and the resource existence may or may not be known. Once the backoff finishes, the XdsClient will reschedule the `respTimer` when it recreates the ADS stream and re-request all the resources.
  - ~TODO: need to check with Mark do we really want to reschedule the initial resource fetch timer for recreated ADS stream. As now the xDS version_info is persistent across ADS streams and the management server is allowed to not send resources that previously sent after the ADS stream is recreated.~
  - __Update__: the initial resource fetch timer should only be scheduled when sending requests for a resource that its existence is not known. Will address this issue in a later PR. This PR will keep the existing behavior.

Watchers can be added to existing `ResourceSubscriber`s. At the time the watcher is added, its callback will be invoked if we've already known the existence of the resource. Otherwise, the watcher will just sit there and wait data or absence to come in the future. 

Every time after processing the xDS response, the extracted and validated resource data (or absence) will be put into each corresponding `ResourceSubscribe`. Then watchers may be invoked.

-------------
Will do the same for LDS/RDS in next PR and replace the existing `ConfigWatcher`.

### CODE REVIEW

**Comment by voidzcy** [comment]
Will conflict with #7461.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 1]
```diff
@@ -1333,6 +1150,131 @@ static ResourceType fromTypeUrl(String typeUrl) {
...
+        return;
+      }
+      boolean oldAbsent = absent;
+      data = null;
+      absent = true;
+      if (!oldAbsent) {
+        for (ResourceWatcher watcher : watchers) {
+          watcher.onResourceDoesNotExist(resource);
+        }
+      }
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl.java:1244]: Is it equivalent to 
```java
if (!absent) {
  absent = true;
  data = null;
  for (ResourceWatcher watcher : watchers) {
    watcher.onResourceDoesNotExist(resource);
  }
}
```
?

**Review by dapengzhang0 — COMMENTED** [review]
LGTM. Some minor comment.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 1 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Yes, fixed.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 2]
```diff
@@ -1333,6 +1150,131 @@ static ResourceType fromTypeUrl(String typeUrl) {
...
+
+    void onData(ResourceUpdate data) {
+      if (respTimer != null && respTimer.isPending()) {
+        respTimer.cancel();
+        respTimer = null;
+      }
+      ResourceUpdate oldData = this.data;
+      this.data = data;
+      absent = false;
+      if (!Objects.equals(oldData, data)) {
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:1231]: Fixed.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java [hunk 3]
```diff
@@ -1025,63 +891,26 @@ private void handleCdsResponse(DiscoveryResponseData cdsResponse) {
...
-          watcher.onChanged(cdsUpdate);
-        }
-      } else if (!clusterNamesToCdsUpdates.containsKey(entry.getKey())
-          && !cdsRespTimers.containsKey(clusterName)) {
-        // Update for previously present resource being removed.
-        for (CdsResourceWatcher watcher : entry.getValue()) {
-          watcher.onResourceDoesNotExist(entry.getKey());
-        }
+    for (String resource : edsResourceSubscribers.keySet()) {
+      ResourceSubscriber subscriber = edsResourceSubscribers.get(resource);
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java]: Fixed.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl.java:1231]: Also `RdsUpdate`.

**Review by dapengzhang0 — APPROVED** [review]

**Pull request merged**